### DISCUSSION TITLE
How to use echart with dark mode?

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#1958] [Ideas / Feature Requests]
### Question

How do you use ui.echart with dark mode?  By default, echart uses dark grey text which is unreadable when your app is in dark mode.  echart background is transparent by default.  So background is fine, text is not.  

# Manual option
One very unappealing option would be to manually set dark mode colors on all text items in all your echarts in the `options` dict.  Then you'd have to dynamically change all those colors in your python code when the app switches to light mode (user toggle).   Not a good solution.

# Themes
echarts has [themes for styling](https://echarts.apache.org/en/download-theme.html), including one for dark mode.  You could download it (dark.js) and manually add it to your app with ui.add_head_html.  But I'm not sure how to dynamically switch between dark theme and light theme.

In js the solution seems to be to [dispose of and rebuild all your charts](https://stackoverflow.com/a/55459118/7663296), passing the chosen theme to echarts.  But nicegui doesn't expose echarts.dispose or echarts.init functions for this to work (nor should nicegui expose them).  This also requires passing all the options including chart data again.

# Solutions?
So what's the solution?  Custom js to rebuild the charts?  I'm concerned this may interfere with nicegui's handling of echart component.  And it requires a reference to the chart `options` with data.  Those are set dynamically in my nicegui app.  Not easy to pass all that to js each time dark mode changes.

Since nicegui handles switching between dark and light mode, it would be great if it also handled updating the echarts theme.  Does anyone have a good solution in the meantime?

**Comment by falkoschindler** [comment]
That's a good question, @ed2050.
I converted this discussion into a feature request, because there's currently no solution.
Maybe someone from the community has an idea how to achieve it?

**Comment by ed2050** [comment]
Thanks @falkoschindler.  Anyone have suggestions?

## Update
I found a partial improvement:

It seems echarts has a [darkMode flag](https://echarts.apache.org/en/option.html#darkMode) in options.  It only [partially works](https://github.com/apache/echarts/issues/18198#issuecomment-1420937979) though.  When set, text color only changes on data labels.  Other text does not change color: title, legend, etc.  Even if `darkMode : true` is set in the subdict for title or textStyle, it has no effect.

I got this limited form of darkMode to work dynamically as follows:
```py
chart = ui.echart (options)

def change_chart_mode () :   
    'toggle dark mode on chart'
    chart.options.update ({ 'darkMode' : not chart.options ['darkMode'] })
    chart.update ()   # render on client

ui.dark_mode (on_change = change_chart_mode)
```

Not a full solution but maybe will lead to something better.
